src/main/java/org/candlepin/util/Util.java

```diff
@@ -702,4 +702,33 @@ public static boolean isFalse(Boolean value) {
         return value == null || Boolean.FALSE.equals(value);
     }
 
+    /**
+     * Checks the provided collection for any duplicate object or null values. Object comparisons are done
+     * using the object's <code>equals</code> implementation.
+     *
+     * @param <T>
+     *  the type of objects to check
+     *
+     * @param collection
+     *  the items to check for duplicates or null values
+     *
+     * @return true if there is a duplicate in the provided collection or a null value, false otherwise
+     */
+    public static <T> boolean containsDuplicateOrNull(Collection<T> collection) {
+        if (collection == null || collection.isEmpty()) {
+            return false;
+        }
+
+        Set<T> objects = new HashSet<>();
+        for (T obj : collection) {
```

Comment:
This may be a fun experiment in perf testing to check if it's faster to do it this way under the assumption that our input *does* contain duplicates or null, or if -- since we're using a set already -- it's faster to dump all of the items into the collection and then check if the set contains null or has fewer elements than our input:

```
Set<T> objects = new HashSet<>(collection);
return objects.contains(null) || objects.size() < collection.size();
```

My guess would be that the iterative method is faster in cases where the input indeed contains dupes or nulls early on, but slower in cases where the input is functionally a set of non-nulls or has many elements with a null or dupe near the tail of the collection's iterator.

Also, if the input collection is a set already, there's no reason to add everything to a new set (iteratively or otherwise). A simple `.contains(null)` check would be sufficient. So our potential solution becomes:

```
Set<T> objects = collection instanceof Set set ? set : new HashSet<>(collection);
return objects.contains(null) || objects.size() < collection.size();
```

---

Something else to consider with this method is whether or not this behavior needs to be generalized or should be handled as part of the conversion. We're doing a lot of rework iterating over the elements just to pass them to another *different* set later, which already imposes these restrictions. That being the case, we could have that conversion do the work of both checks here:

```
NonNullLinkedHashSet<String> envSet = new NonNullLinkedHashSet<>();

try {
  envSet.addAll(environmentIds);

  if (envSet.size() < environmentIds.size()) {
    throw new BadRequestException(this.i18n.tr("Request contains duplicate environment IDs"));
  }
}
catch (IllegalArgumentException e) {
  throw new BadRequestException(this.i18n.tr("Environment IDs cannot be null or empty"));
}
```

Reply:
That is a great point, I really like adding the environment IDs into the `NonNullLinkedHashSet` and determining if there were duplicates by comparing the size with the provided list! I updated `OwnerResource` to use this approach and removed the `Util.containsDuplicateOrNull` method